packages/integration-sdk-entity-validator/src/validator.ts

```diff
@@ -0,0 +1,158 @@
+import Ajv, { AnySchema } from 'ajv';
+import addFormats from 'ajv-formats';
+import { addJ1Formats } from './j1Formats';
+import {
+  ajvErrorToEntityValidationError,
+  unknownPropertySymbol,
+  isEntityValidationError,
+  EntityValidationError,
+} from './entityValidationError';
+import { assertEntity } from './assertEntity';
+
+type ValidateEntityResult = {
+  isValid: boolean;
+  errors: EntityValidationError[] | null;
+  skippedSchemas: SkippedSchema[] | null;
+};
+
+type SkippedSchema = {
+  schemaId: string;
+  reason: 'not-found' | 'type-already-validated';
+  type: 'type' | 'class';
+};
+
+type ValidateEntityOptions = {
+  forceClassValidationWithValidatedType?: boolean;
+};
+
+const convertUnknownErrorToEntityValidationError = (
+  error: unknown,
+): EntityValidationError => {
+  if (isEntityValidationError(error)) {
+    return error;
+  }
+
+  if (error instanceof Error) {
+    return {
+      schemaId: 'unknown',
+      message: error.message,
+      property: unknownPropertySymbol,
+      validation: 'unknown',
+    };
+  }
+
+  return {
+    schemaId: 'unknown',
+    message: 'Unknown error',
+    property: unknownPropertySymbol,
+    validation: 'unknown',
+  };
+};
+
+export class EntityValidator {
+  private ajvInstance: Ajv;
+
+  constructor({ schemas }: { schemas?: AnySchema[] }) {
+    this.ajvInstance = addJ1Formats(
+      addFormats(
+        new Ajv({
+          strictSchema: false,
+          allErrors: true,
+        }),
+      ),
+    );
+
+    if (schemas) {
+      this.addSchemas(schemas);
+    }
+  }
+
+  addSchemas(schema: AnySchema | AnySchema[]) {
+    this.ajvInstance.addSchema(schema);
+    return this;
+  }
+
+  validateEntity(
+    entity: object,
+    {
+      forceClassValidationWithValidatedType = false,
+    }: ValidateEntityOptions = {},
+  ): ValidateEntityResult {
+    const errors: EntityValidationError[] = [];
+    const skippedSchemas: SkippedSchema[] = [];
+
+    try {
+      assertEntity(entity);
```

Comment:
I don't know if it's possible, but I'll say it anyway and leave it to you to decide. The nesting is making the code hard to follow, so if possible, it would be better to flatten out this function a little bit.

Reply:
Refactored. Everything is inside a single for loop right now making the code clear.

Reply:
It's beautiful!